File: templates/ObjectiveC-UIKit/Core/Extensions/UIView+Extensions.m
//
//  UIView+Extensions.m
//  ObjectiveC-UIKit
//

#import "UIView+Extensions.h"

@implementation UIView (Extensions)

- (NSArray<NSLayoutConstraint *> *)pinEdgesToView:(UIView *)view {
    return [self pinEdgesToView:view insets:UIEdgeInsetsZero];
}

- (NSArray<NSLayoutConstraint *> *)pinEdgesToView:(UIView *)view
                                          insets:(UIEdgeInsets)insets {
    self.translatesAutoresizingMaskIntoConstraints = NO;
    NSArray<NSLayoutConstraint *> *constraints = @[
        [self.topAnchor constraintEqualToAnchor:view.topAnchor
                                       constant:insets.top],
        [self.leadingAnchor constraintEqualToAnchor:view.leadingAnchor
                                           constant:insets.left],
        [self.trailingAnchor constraintEqualToAnchor:view.trailingAnchor
                                            constant:-insets.right],
        [self.bottomAnchor constraintEqualToAnchor:view.bottomAnchor
                                          constant:-insets.bottom],
    ];
    [NSLayoutConstraint activateConstraints:constraints];
    return constraints;
}

- (NSArray<NSLayoutConstraint *> *)pinEdgesToSafeAreaOfView:(UIView *)view {
    self.translatesAutoresizingMaskIntoConstraints = NO;
    UILayoutGuide *guide = view.safeAreaLayoutGuide;
    NSArray<NSLayoutConstraint *> *constraints = @[
        [self.topAnchor constraintEqualToAnchor:guide.topAnchor],
        [self.leadingAnchor constraintEqualToAnchor:guide.leadingAnchor],
        [self.trailingAnchor constraintEqualToAnchor:guide.trailingAnchor],
        [self.bottomAnchor constraintEqualToAnchor:guide.bottomAnchor],
    ];
    [NSLayoutConstraint activateConstraints:constraints];
    return constraints;
}

- (NSArray<NSLayoutConstraint *> *)centerInView:(UIView *)view {
    self.translatesAutoresizingMaskIntoConstraints = NO;
    NSArray<NSLayoutConstraint *> *constraints = @[
        [self.centerXAnchor constraintEqualToAnchor:view.centerXAnchor],
        [self.centerYAnchor constraintEqualToAnchor:view.centerYAnchor],
    ];
    [NSLayoutConstraint activateConstraints:constraints];
    return constraints;
}

@end

File: templates/Swift-UIKit/Core/Extensions/UIView+Extensions.swift
import UIKit

extension UIView {

    /// Adds `subview` and pins all four edges to the receiver, optionally insetting.
    /// Disables `translatesAutoresizingMaskIntoConstraints` on the subview.
    func addPinned(_ subview: UIView, insets: UIEdgeInsets = .zero) {
        addSubview(subview)
        subview.pinEdges(to: self, insets: insets)
    }

    /// Pins the receiver's edges to another view.
    func pinEdges(to other: UIView, insets: UIEdgeInsets = .zero) {
        translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            topAnchor.constraint(equalTo: other.topAnchor, constant: insets.top),
            leadingAnchor.constraint(equalTo: other.leadingAnchor, constant: insets.left),
            trailingAnchor.constraint(equalTo: other.trailingAnchor, constant: -insets.right),
            bottomAnchor.constraint(equalTo: other.bottomAnchor, constant: -insets.bottom),
        ])
    }

    /// Pins the receiver's edges to another view's safe area layout guide.
    func pinEdgesToSafeArea(of other: UIView, insets: UIEdgeInsets = .zero) {
        translatesAutoresizingMaskIntoConstraints = false
        let guide = other.safeAreaLayoutGuide
        NSLayoutConstraint.activate([
            topAnchor.constraint(equalTo: guide.topAnchor, constant: insets.top),
            leadingAnchor.constraint(equalTo: guide.leadingAnchor, constant: insets.left),
            trailingAnchor.constraint(equalTo: guide.trailingAnchor, constant: -insets.right),
            bottomAnchor.constraint(equalTo: guide.bottomAnchor, constant: -insets.bottom),
        ])
    }

    /// Centers the receiver within another view.
    func center(in other: UIView) {
        translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            centerXAnchor.constraint(equalTo: other.centerXAnchor),
            centerYAnchor.constraint(equalTo: other.centerYAnchor),
        ])
    }

    /// Constrains the receiver to a fixed size.
    func constrainSize(width: CGFloat? = nil, height: CGFloat? = nil) {
        translatesAutoresizingMaskIntoConstraints = false
        if let width { widthAnchor.constraint(equalToConstant: width).isActive = true }
        if let height { heightAnchor.constraint(equalToConstant: height).isActive = true }
    }
}
